### DISCUSSION TITLE
Welcome to staticSearch Discussions!

### DISCUSSION THREAD

**Discussion by martindholmes** [projectEndings/staticSearch#237] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other users and potential users of staticSearch. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas and pitch suggestions more informally than you might do when raising an issue.
  * Engage with other community members.
  * Welcome others and are open-minded. 

  To get started, comment below with an introduction of yourself and tell us about what you use staticSearch for, or what you're thinking about using it for.

**Comment by mprell** [comment]
First of all, many thanks for the staticSearch and its continuous maintenance! In my opinion, there is currently nothing comparable that fits so well with X technologies and continues to be maintained. And staticSearch seems to me to be a great contribution not only to longevity but also to the greening of DH projects.
I'm Martin and I have implemented staticSearch in a website for the presentation of diaries of sea voyages of the Moravian Church in the 18th century. 
Search site: https://www.moravians-at-sea.de/suche.html
Data and configuration file etc.: https://github.com/mprell/moravians-at-sea-Prod
I would like to ask two questions:
1. Is there already experience, what amount of data (e.g. the number of json files in the stems folder or the size of the indexed static xhtml files) noticeably slow down the performance of StaticSearch? I am asking in order to be able to estimate whether staticSearch is also suitable for projects with 100,000 and more letters and diary entries or something similar. It is difficult to find out what quantity of data the reference projects mentioned [here](https://endings.uvic.ca/staticSearch/docs/projectsUsingSS.html) use.
2. Does anyone know of a project that integrates a German stemmer or is the implementation of a German stemmer possibly even in progress? So far I have been using wildcard searches.

**Reply by joeytakeda** [reply]
Hi @mprell ! Thanks for the kind words and support :-) Great to see your project too -- we'd be glad to list it on the "Projects Using StaticSearch" page you link, if that's OK with you?

To answer your questions:

> Is there already experience, what amount of data (e.g. the number of json files in the stems folder or the size of the indexed static xhtml files) noticeably slow down the performance of StaticSearch? I am asking in order to be able to estimate whether staticSearch is also suitable for projects with 100,000 and more letters and diary entries or something similar. It is difficult to find out what quantity of data the reference projects mentioned [here](https://endings.uvic.ca/staticSearch/docs/projectsUsingSS.html) use.

We have some better stats and figures here for document collections: https://www.balisage.net/Proceedings/vol27/html/Takeda01/BalisageVol27-Takeda01.html#statistics 

In terms of front-end: With lots of documents, there can be a slight delay in the initial start-up of staticSearch since it loads in all of the titles on initialization. Otherwise, though, there's no performance hit per search when the size of site increases. staticSearch only searches for the string that was searched, so, in other words, the front-end JS doesn't know or care about the rest of the stems, just the ones it is looking for. 

In terms of back-end: lots of documents can slow down the indexing processing. You may need to give `ant` additional memory, depending on the size of the document collection (see #323 and #157 for more details). 

> Does anyone know of a project that integrates a German stemmer or is the implementation of a German stemmer possibly even in progress? So far I have been using wildcard searches

I'm not aware of a German implementation in progress, but since there is a Porter implementation, this is something that could be plugged in: 

http://snowball.tartarus.org/algorithms/german/stemmer.html

It looks like there's already one in JS (https://github.com/klaemo/snowball-german/blob/master/index.js), so this would just need to be implemented in XSLT as well. 

We do have some "blue sky" plans for ways of adopting all of the Porter stemmers in one go (#244 ), but that's been put a bit on the back-burner while we're working toward a new major release.

**Reply by mprell** [reply]
Hi @joeytakeda
It would be a great pleasure to see my project in the staticSearch projects list. Thank you!
Many thanks for the hint to your Balisage article, as well; I didn't know it yet. This allows me to better assess the performance effects of large amounts of data on the front-end.
I'll take a closer look at the Porter Stemmer for German, many thanks!

**Comment by mprell** [comment]
Hi Joey and Martin,
I am currently testing staticSearch for over 53,000 html files of the Goethe Biographica project (averaging 20,5 KB each) and 100,000+ created JSON-Files for the staticSearch.
To speed up the build process a bit, I modified the `build.xml` so I can choose whether to tokenize all html files during every build or only those that have changed since the last build (so called incremental build); here is the relevant commit: https://github.com/mprell/staticSearch/commit/5bed3721fcb9b2b62d93af239eaa6fa60b0a10af
Then I can launch the incremental build by runing the `incremental`-target . And to ensure that already tokenized files are accessed during a new build, the `ssTemp` folder must not be deleted after a build, so I removed the `clean` step: https://github.com/mprell/staticSearch/commit/0c53d35dafac7daf79f4e32bbf23d4a2a75fa074
So far, everything seems to be working as expected. But does it really act like a completely new build as long as the JSON files are regenerated from scratch? Or am I overlooking something? I would appreciate any hint on speeding up the build process — beyond decreasing the HTML nodes to be indexed via `rules` in config.xml.
Thank you very much and many greetings,
Martin
